File: yacoo_rpg/src/game/run.test.ts
import { describe, expect, it } from 'vitest';
import { createDefaultMeta } from './equipment';
import {
  advanceChapter,
  advanceNode,
  createChapterMap,
  createRun,
  currentNode,
  isChapterCleared,
  isRunOver,
  settleRun
} from './run';

describe('createChapterMap', () => {
  it('creates a map with the correct node count', () => {
    const map = createChapterMap(1, () => 0.5);
    expect(map.nodes).toHaveLength(6);
  });

  it('always starts with battle and ends with boss', () => {
    const map = createChapterMap(1, Math.random);
    expect(map.nodes[0].type).toBe('battle');
    expect(map.nodes[5].type).toBe('boss');
  });

  it('is deterministic given the same rng sequence', () => {
    let seed = 0;
    const seededRng = () => ((seed = (seed * 16807 + 0) % 2147483647) / 2147483647);
    const a = createChapterMap(1, seededRng);
    seed = 0;
    const seededRng2 = () => ((seed = (seed * 16807 + 0) % 2147483647) / 2147483647);
    const b = createChapterMap(1, seededRng2);
    expect(a.nodes.map((n) => n.type)).toEqual(b.nodes.map((n) => n.type));
  });

  it('starts with all nodes uncleared', () => {
    const map = createChapterMap(2, () => 0);
    expect(map.nodes.every((n) => !n.cleared)).toBe(true);
  });
});

describe('createRun', () => {
  it('creates a run with starting stats from equipment', () => {
    const meta = createDefaultMeta();
    const run = createRun(meta, () => 0.5);
    // Default equipment: maxHp = 120 + 1*15 = 135
    expect(run.hp).toBe(135);
    expect(run.maxHp).toBe(135);
    expect(run.chapter).toBe(1);
    expect(run.nodeIndex).toBe(0);
    expect(run.diceCount).toBe(5);
    expect(run.maxRolls).toBe(3);
  });
});

describe('advanceNode', () => {
  it('marks the current node as cleared and increments nodeIndex', () => {
    const meta = createDefaultMeta();
    const run = createRun(meta, () => 0.5);
    const next = advanceNode(run);
    expect(next.map.nodes[0].cleared).toBe(true);
    expect(next.nodeIndex).toBe(1);
    expect(run.nodeIndex).toBe(0); // original unchanged (immutable)
  });
});

describe('isRunOver / isChapterCleared', () => {
  it('detects run over when hp reaches 0', () => {
    const meta = createDefaultMeta();
    const run = createRun(meta, () => 0.5);
    expect(isRunOver(run)).toBe(false);
    expect(isRunOver({ ...run, hp: 0 })).toBe(true);
  });

  it('detects chapter cleared when nodeIndex exceeds map length', () => {
    const meta = createDefaultMeta();
    const run = createRun(meta, () => 0.5);
    expect(isChapterCleared(run)).toBe(false);
    expect(isChapterCleared({ ...run, nodeIndex: 6 })).toBe(true);
  });
});

describe('currentNode', () => {
  it('returns the node at nodeIndex', () => {
    const meta = createDefaultMeta();
    const run = createRun(meta, () => 0.5);
    const node = currentNode(run);
    expect(node).not.toBeNull();
    expect(node?.type).toBe('battle');
  });

  it('returns null when past the end', () => {
    const meta = createDefaultMeta();
    const run = { ...createRun(meta, () => 0.5), nodeIndex: 999 };
    expect(currentNode(run)).toBeNull();
  });
});

describe('advanceChapter', () => {
  it('increments chapter and resets nodeIndex', () => {
    const meta = createDefaultMeta();
    const run = createRun(meta, () => 0.5);
    const next = advanceChapter(run, () => 0.3);
    expect(next.chapter).toBe(2);
    expect(next.nodeIndex).toBe(0);
    expect(next.map.chapter).toBe(2);
  });
});

describe('settleRun', () => {
  it('adds coins and updates bestChapter', () => {
    const meta = createDefaultMeta();
    const run = createRun(meta, () => 0.5);
    const run5 = { ...run, chapter: 5 };
    const settled = settleRun(meta, run5, 120);
    expect(settled.coins).toBe(120);
    expect(settled.bestChapter).toBe(5);
    expect(settled.totalRuns).toBe(1);
    expect(settled.runInProgress).toBeUndefined();
  });

  it('does not downgrade bestChapter', () => {
    const meta = { ...createDefaultMeta(), bestChapter: 10 };
    const run = createRun(meta, () => 0.5);
    const settled = settleRun(meta, { ...run, chapter: 3 }, 0);
    expect(settled.bestChapter).toBe(10);
  });
});

File: yacoo_rpg-android/app/src/test/java/com/yacoo/rpg/game/RunTest.kt
package com.yacoo.rpg.game

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class RunTest {

    @Test fun `createChapterMap produces six nodes with boss last`() {
        val map = createChapterMap(1) { 0.5 }
        assertEquals(6, map.nodes.size)
        assertEquals(NodeType.BATTLE, map.nodes.first().type)
        assertEquals(NodeType.BOSS, map.nodes.last().type)
    }

    @Test fun `createChapterMap keeps first and last fixed under shuffle`() {
        listOf(0.0, 0.25, 0.5, 0.75, 0.99).forEach { seed ->
            val map = createChapterMap(2) { seed }
            assertEquals(NodeType.BATTLE, map.nodes.first().type)
            assertEquals(NodeType.BOSS, map.nodes.last().type)
        }
    }

    @Test fun `advanceNode marks current cleared and increments index`() {
        val run = createRun(createDefaultMeta()) { 0.5 }
        val advanced = advanceNode(run)
        assertEquals(run.nodeIndex + 1, advanced.nodeIndex)
        assertTrue(advanced.map.nodes[run.nodeIndex].cleared)
    }

    @Test fun `isChapterCleared true when index past end`() {
        val run = createRun(createDefaultMeta()) { 0.5 }
        val past = run.copy(nodeIndex = run.map.nodes.size)
        assertTrue(isChapterCleared(past))
        assertFalse(isChapterCleared(run.copy(nodeIndex = run.map.nodes.size - 1)))
    }

    @Test fun `advanceChapter produces chapter+1 with fresh map and reset index`() {
        val run = createRun(createDefaultMeta()) { 0.5 }
        val advanced = advanceChapter(run) { 0.5 }
        assertEquals(run.chapter + 1, advanced.chapter)
        assertEquals(0, advanced.nodeIndex)
        assertEquals(6, advanced.map.nodes.size)
        assertNotSame(run.map, advanced.map)
    }

    @Test fun `applyHealToRun caps at maxHp`() {
        val run = createRun(createDefaultMeta()) { 0.5 }.copy(hp = 100, maxHp = 120)
        assertEquals(120, applyHealToRun(run, 50).hp)
    }

    @Test fun `applyHealToRun stacks below maxHp`() {
        val run = createRun(createDefaultMeta()) { 0.5 }.copy(hp = 30, maxHp = 120)
        assertEquals(60, applyHealToRun(run, 30).hp)
    }

    @Test fun `currentNode returns null past end`() {
        val run = createRun(createDefaultMeta()) { 0.5 }
        val past = run.copy(nodeIndex = run.map.nodes.size)
        assertNull(currentNode(past))
    }

    @Test fun `settleRun increments bestChapter when run is further`() {
        val meta = createDefaultMeta().copy(bestChapter = 1)
        val run  = createRun(meta) { 0.5 }.copy(chapter = 3)
        val settled = settleRun(meta, run, coinsEarned = 0)
        assertEquals(3, settled.bestChapter)
        assertEquals(1, settled.totalRuns)
        assertNull(settled.runInProgress)
        assertNull(settled.lastCombatResult)
    }
}
